### PR TITLE
Update to upstream Python 3.10.11

### PR DESCRIPTION
**Pull request by andreittr**
This change set updates the unikraft Python port to 3.10.11, along with necessary compatibility patches.

This is part of a larger work and depends on several other PRs:
- [x] https://github.com/unikraft/lib-libffi/pull/1
- [x] https://github.com/unikraft/lib-libuuid/pull/4
- [x] https://github.com/unikraft/unikraft/pull/996
- [x] https://github.com/unikraft/unikraft/pull/994
- [x] https://github.com/unikraft/lib-compiler-rt/pull/14
- [x] https://github.com/unikraft/lib-compiler-rt/pull/16
- [x] https://github.com/unikraft/unikraft/pull/953
- [x] https://github.com/unikraft/unikraft/pull/952
- [x] https://github.com/unikraft/unikraft/pull/951
- [x] https://github.com/unikraft/unikraft/pull/936
- [x] https://github.com/unikraft/lib-sqlite/pull/6
- [x] https://github.com/unikraft/lib-openssl/pull/5
- [x] https://github.com/unikraft/lib-musl/pull/55
- [x] https://github.com/unikraft/unikraft/pull/924

Tested by running python's own test suite. Tests _usually_ pass, with some functionality like fork/exec unavailable.
Built with GCC & Clang with the following libraries:
```
LIBS-y := $(LIBS-y):$(UK_LIBS)/lib-compiler-rt
LIBS-y := $(LIBS-y):$(UK_LIBS)/lib-musl
LIBS-y := $(LIBS-y):$(UK_LIBS)/lib-libffi
LIBS-y := $(LIBS-y):$(UK_LIBS)/lib-lwip
LIBS-y := $(LIBS-y):$(UK_LIBS)/lib-bzip2
LIBS-y := $(LIBS-y):$(UK_LIBS)/lib-sqlite
LIBS-y := $(LIBS-y):$(UK_LIBS)/lib-openssl
LIBS-y := $(LIBS-y):$(UK_LIBS)/lib-libuuid
LIBS-y := $(LIBS-y):$(UK_LIBS)/lib-zlib
LIBS-y := $(LIBS-y):$(UK_LIBS)/lib-python3
```

### CODE REVIEW

File: Config.uk [hunk 1]
```diff
@@ -15,20 +15,20 @@ menuconfig LIBPYTHON3
...
 	select LIBPOSIX_SOCKET
 	select LIBPOSIX_EVENT
-	select LIBLWIP
-	select LWIP_DHCP
-	select LWIP_DNS
-	select LWIP_IPV6
+	imply LIBLWIP
+	imply LWIP_DHCP
+	imply LWIP_DNS
+	imply LWIP_IPV6
```
Comment (StefanJum) [inline Config.uk:21]: Is there a reason to imply instead of select?

File: Makefile.uk [hunk 2]
```diff
@@ -61,22 +67,32 @@ LIBPYTHON3_SRC = $(LIBPYTHON3_ORIGIN)/$(LIBPYTHON3_BASENAME)
...
+LIBPYTHON3_SUPPRESS_FLAGS-y += -Wno-missing-field-initializers
+LIBPYTHON3_SUPPRESS_FLAGS-y += -Wno-implicit-fallthrough
+LIBPYTHON3_SUPPRESS_FLAGS-$(call clang_version_ge,13,0) += -Wno-cast-function-type
+LIBPYTHON3_SUPPRESS_FLAGS-$(call have_gcc) += -Wno-stringop-truncation
+LIBPYTHON3_SUPPRESS_FLAGS-y += -Wno-char-subscripts
+LIBPYTHON3_SUPPRESS_FLAGS-y += -Wno-sign-compare
+LIBPYTHON3_SUPPRESS_FLAGS-$(call have_gcc) += -Wno-maybe-uninitialized
+LIBPYTHON3_SUPPRESS_FLAGS-$(call have_clang) += -Wno-documentation-unknown-command
+LIBPYTHON3_SUPRESS_CFLAGS-y   += $(LIBPYTHON3_SUPPRESS_FLAGS-y) -Wno-pointer-to-int-cast -Wno-int-to-pointer-cast
+LIBPYTHON3_SUPRESS_CXXFLAGS-y += $(LIBPYTHON3_SUPPRESS_FLAGS-y)
```
Comment (StefanJum) [inline Makefile.uk]: I would allign these all to `+=` or / and do more than one flag per line.

File: patches/0005-Replace-getdents64-syscall-with-libc-call.patch [hunk 3]
```diff
@@ -0,0 +1,26 @@
+From f607d18e9ee479dd30d4c1f737c08f2d7642122a Mon Sep 17 00:00:00 2001
+From: Andrei Tatar <andrei@unikraft.io>
+Date: Mon, 10 Jul 2023 16:23:27 +0200
+Subject: [PATCH 5/5] Replace getdents64 syscall with libc call
```
Comment (StefanJum) [inline patches/0005-Replace-getdents64-syscall-with-libc-call.patch]: Why is this needed?

**Review by StefanJum — CHANGES_REQUESTED** [review]
Thanks @andreittr. Besides the few comments:

* When I try to build with the libraries list you provided an with no extensions enabled for libpython, I get the following build error:
```
/usr/bin/ld: workdir/apps/python3/build/libffi.o: in function `ffi_prep_cif_var':
workdir/apps/python3/build/libffi/origin/libffi-3.4.4/src/prep_cif.c:236: undefined reference to `ffi_type_sint32'

```

If the build depends on the python extensions being enabled, maybe we can check that in the `Config.uk` of the libraries (`libffi`), or enable the exensions by default.

* If I build with the extensions enabled it builds fine, but it fails at runtime with the following error:
```
Python path configuration:
  PYTHONHOME = '/lib/python3.10'
  PYTHONPATH = '/lib/python3.10'
  program name = 'build/python3_qemu-x86_64'
  isolated = 0
  environment = 1
  user site = 1
  import site = 1
  sys._base_executable = '/build/python3_qemu-x86_64'
  sys.base_prefix = '/lib/python3.10'
  sys.base_exec_prefix = '/lib/python3.10'
  sys.platlibdir = 'lib'
  sys.executable = '/build/python3_qemu-x86_64'
  sys.prefix = '/lib/python3.10'
  sys.exec_prefix = '/lib/python3.10'
  sys.path = [
    '/lib/python3.10',
    '/lib/python3.10/lib/python310.zip',
    '/lib/python3.10',
    '/lib/python3.10/lib/python3.10/lib-dynload',
  ]
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'
```
It works if I change the filesystem and replace all the `3.7` entries to `3.10`, so the `rootfs.tar.gz` (in the `app-python3` repository) should also be updated.

File: Config.uk (hunk 1 above)
Comment (andreittr) [reply Config.uk:21]: Yes, so one can select an alternative network stack without hacking the Config.uk file, while keeping a sane default.

File: Makefile.uk (hunk 2 above)
Comment (andreittr) [reply Makefile.uk]: I'd avoid more than one flag per line unless all flags act as an inseparable whole, otherwise it will make future diffs more unwieldy.
Adding large amounts of whitespace only to align long lines on the assignment operator also looks ugly IMO, so I think cleanest would be to group the conditional flags at the end.

File: patches/0005-Replace-getdents64-syscall-with-libc-call.patch (hunk 3 above)
Comment (andreittr) [reply patches/0005-Replace-getdents64-syscall-with-libc-call.patch]: The patch itself? To be honest I no longer remember, it's been a long time. I did some basic tests now with the two syscall replacing patches removed and it seems ok, so must've been a fix for an issue that's no longer relevant. I'll remove the patches but keep them on hand in case issues spring back up.

**Comment by andreittr** [comment]
>     * When I try to build with the libraries list you provided an with no extensions enabled for libpython, I get the following build error:

I believe that is due to building libffi (without the python extension requiring it => it's not getting used) but not enabling its option to provide the `ffi_type` functions; the reason is that either libffi or python's ctypes defines these functions, and by default libffi doesn't provide them itself.
I think this can be made clearer in libffi's Config.uk (make condition negative; have python's ctypes select said condition); will update libffi's PR with that.

>     * If I build with the extensions enabled it builds fine, but it fails at runtime with the following error:

`app-python-3` has not been updated _at all_ for py3.10 (there's no PR for it yet) and the prebuilt rootfs is evidently out of date and could never work. A more extensive update to that repo is in the works.
For the time being please test rebuilding the rootfs with the new version.

**Comment by StefanJum** [comment]
> will update libffi's PR with that.

Perfect, thanks.


> `app-python-3` has not been updated _at all_ for py3.10 (there's no PR for it yet) and the prebuilt rootfs is evidently out of date and could never work. A more extensive update to that repo is in the works. For the time being please test rebuilding the rootfs with the new version.

Yes, I did test using an updated rootfs and everything works fine :rocket: , but I think the updates on `app-python3` should be opened and merged as part of this PR set, to keep stuff working.

**Comment by andreittr** [comment]
> Yes, I did test using an updated rootfs and everything works fine rocket , but I think the updates on `app-python3` should be opened and merged as part of this PR set, to keep stuff working.

I agree, app PR is coming soon.

**Comment by andreittr** [comment]
App update up: https://github.com/unikraft/app-python3/pull/16

**Review by StefanJum — APPROVED** [review]
All good, thank you @andreittr :heart: 
I'll add the tag after the discussions on [`libuuid`](https://github.com/unikraft/lib-libuuid/pull/4) and [`lib-sqlite`](https://github.com/unikraft/lib-sqlite/pull/6) (on throwing build-time errors instead of using `depends on`) are solved, so they can be merged before this.

**Review by mariasfiraiala — APPROVED** [review]
All good on my side, thanks a lot for the consistent work and attention @andreittr @StefanJum!

Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

**Review by StefanJum — APPROVED** [review]
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
